### PR TITLE
Feature/full type safety

### PR DESCRIPTION
**Pull request by AlexKiz**

### CODE REVIEW

File: client/src/components/Admin/cities.list/CitiesList.tsx [hunk 1]
```diff
@@ -1,18 +1,21 @@
 import axios from "axios";
-import { useState, useEffect } from "react";
+import { useState, useEffect, FC } from "react";
 import {Link} from 'react-router-dom'
 import '../cities.list/cities-list.css'
+import { City } from '../../../types/types'
 
-const CitiesList = () => {
+interface CitiesListProps {}
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/cities.list/CitiesList.tsx:7]: Maybe we need to put all interfaces in a separate file if we are using them in many places. If we're using the interface in the one file let's put it near this file.
https://www.tatvasoft.com/blog/reactjs-best-practices/

File: client/src/components/Admin/cities.list/CitiesList.tsx [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
+
+    const [cities, setCities] = useState<City[]>([])
 
 
     useEffect(()=> {
 
         const readAllCities = async () => {
 
-            const {data} = await axios.get(`/city`)
+            const {data} = await axios.get<City[]>(`/city`)
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/cities.list/CitiesList.tsx:18]: Constant for the '/city'

File: client/src/components/Admin/cities.list/CitiesList.tsx [hunk 3]
```diff
@@ -56,10 +59,10 @@ const CitiesList = () => {
                     {
                         cities.map((elem) => (
                             <tr>
-                                <td>{`${elem.id}`}</td>
-                                <td>{`${elem.name}`}</td>
-                                <button className='button-update'><Link to = {`/admin/city-controller/${elem.id}/${elem.name}`}>Update</Link></button>
-                                <button className='button-delete' onClick = {() => {onDelete(elem.id)}}>Delete</button>
+                                <td>{`${elem.cityId}`}</td>
+                                <td>{`${elem.cityName}`}</td>
+                                <button className='button-update'><Link to = {`/admin/city-controller/${elem.cityId}/${elem.cityName}`}>Update</Link></button>
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/cities.list/CitiesList.tsx:64]: Constant for route

File: client/src/components/Admin/master.controller/MasterController.tsx [hunk 4]
```diff
@@ -29,21 +32,21 @@ const MasterController = () => {
                     const currentMasterCities = currentMaster[0].cities.map((elem) => {return elem.cityId})
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/master.controller/MasterController.tsx:32]: What if we will have currentMaster === []?

File: client/src/components/Admin/order.controller/OrderController.tsx [hunk 5]
```diff
@@ -1,47 +1,49 @@
...
+import { Params, User, Clock, City, Master } from '../../../types/types'
 
 const currentDate = new Date() 
 const currentDay = (currentDate.getDate() < 10) ? `0${currentDate.getDate()}` : currentDate.getDate()
 const currentMonth = ((currentDate.getMonth() + 1) < 10) ? `0${(currentDate.getMonth() + 1)}` : (currentDate.getMonth() + 1)
 const currentYear = currentDate.getFullYear()
 const today = `${currentYear}-${currentMonth}-${currentDay}`
 
-const openingHours = ['09:00', '10:00', '11:00', '12:00', '13:00', '14:00', '15:00', '16:00', '17:00', '18:00']
+const openingHours: string[] = ['09:00', '10:00', '11:00', '12:00', '13:00', '14:00', '15:00', '16:00', '17:00', '18:00']
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/order.controller/OrderController.tsx:13]: Let's move the constants to separate files

File: client/src/components/Admin/order.controller/OrderController.tsx [hunk 6]
```diff
@@ -1,47 +1,49 @@
...
+    const [masterId, setMasterId] = useState<number>(0)
+    const [masters, setMasters] = useState<Master[]>([])
 
 
     useEffect(() => {
         
         const readAllUsers = async () => {
 
-            const {data} = await axios.get(`/user`)
+            const {data} = await axios.get<User[]>(`/user`)
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/order.controller/OrderController.tsx:43]: constant

File: client/src/components/Admin/order.controller/OrderController.tsx [hunk 7]
```diff
@@ -1,47 +1,49 @@
...
 
             setUsers(data)
-            setUserId(propsUserId)
+            setUserId(+propsUserId)
```
Comment (oleshkoo-clockwise) [inline client/src/components/Admin/order.controller/OrderController.tsx:46]: Why do you need plus sign here?

File: client/src/types/types.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
+export type Params = {
+    propsOrderId: string
```
Comment (oleshkoo-clockwise) [inline client/src/types/types.ts:2]: why do you need prefix "props"?

File: client/src/types/types.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    propsCityName: string
+    propsOrderDate: string
+    propsOrderTime: string
+    propsMasterId: string
+    propsMasterName: string
+    ratingIdentificator: string
+}
+
+export type City = {
+    cityId: number
```
Comment (oleshkoo-clockwise) [inline client/src/types/types.ts:17]: id or cityId?

File: src/controller/auth.controller.ts [hunk 10]
```diff
@@ -14,7 +16,7 @@ const Auth = async (req, res) => {
 
         if(await bcrypt.compare(adminPassword,hashPass)) {
```
Comment (oleshkoo-clockwise) [inline src/controller/auth.controller.ts:18]: remove all useless enters

File: src/controller/order.controller.ts [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+export const putRatedOrder = async (req: Request, res: Response) => {
+
+    try {
+
+        const {id, order_rated, master_id} = req.body
+        
+        const readMasterRating = await db.query('SELECT rated_sum, rated_quantity FROM masters WHERE id = $1', [master_id])
+
```
Comment (oleshkoo-clockwise) [inline src/controller/order.controller.ts:113]: incorrect master id and many errors, the same stuff for other incorrect fields

File: src/routes/admin.router.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+import { getOrder, putOrder, deleteOrder } from '../controller/order.controller'
+import { putOrderValidate, deleteOrderValidate } from '../controller/order.validate'
+
+import { getUser, putUser, deleteUser } from '../controller/user.controller'
+import { putUserValidate, deleteUserValidate } from '../controller/user.validate'
+
+const router = Router()
+
+router.get('/city', [isAuth], getCity)
```
Comment (oleshkoo-clockwise) [inline src/routes/admin.router.ts:18]: constants

**Review by oleshkoo-clockwise — CHANGES_REQUESTED** [review]

File: client/src/components/Admin/order.controller/OrderController.tsx (hunk 7 above)
Comment (AlexKiz) [reply client/src/components/Admin/order.controller/OrderController.tsx:46]: > 

TS complaine about type string of 'propsUserId' that useParams returns

Comment (AlexKiz) [reply client/src/components/Admin/order.controller/OrderController.tsx:46]: @oleshkoo-clockwise  TS complaine about type string of 'propsUserId' that useParams returns

**Review by oleshkoo-clockwise — APPROVED** [review]

**Pull request merged**